### PR TITLE
PHOENIX-6475 Use org.openlabtesting.protobuf:protoc when building on Linux ARM64

### PR DESCRIPTION
**Pull request by martin-g**
Stock Protobuf-Java adds aarch64 protoc binary in 3.5.0.
To be able to build on Linux ARM64 with Protobuf 2.x we should use https://github.com/os72/protoc-jar:2.6.1+

### CODE REVIEW

File: phoenix-pherf/pom.xml [hunk 1]
```diff
@@ -133,10 +133,6 @@
       <groupId>org.apache.phoenix.thirdparty</groupId>
       <artifactId>phoenix-shaded-commons-cli</artifactId>
     </dependency>
-    <dependency>
-      <groupId>commons-io</groupId>
-      <artifactId>commons-io</artifactId>
-    </dependency>
```
Comment (martin-g) [inline phoenix-pherf/pom.xml:139]: This dependency is removed because it is a duplicate of another one at line 125.
Let me know if it should be in a separate PR!

File: pom.xml [hunk 2]
```diff
@@ -107,7 +107,10 @@
     <log4j.version>1.2.17</log4j.version>
     <disruptor.version>3.3.6</disruptor.version>
     <slf4j.version>1.7.25</slf4j.version>
-    <protobuf-java.version>2.5.0</protobuf-java.version>
+    <!-- com.google repo will be used except on Aarch64 platform. -->
+    <protobuf.group>com.google.protobuf</protobuf.group>
+    <protobuf-java.version>2.6.1</protobuf-java.version>
```
Comment (martin-g) [inline pom.xml:112]: I've updated to 2.6.1 to keep the x86_64 and aarch64 versions in sync.

File: phoenix-pherf/pom.xml (hunk 1 above)
Comment (stoty) [reply phoenix-pherf/pom.xml:139]: Yes, please open a separate JIRA for this.

**Review by stoty — CHANGES_REQUESTED** [review]
This patch changes too much, and we specifically want to use the exact same version (2.5.0) that HBase and Hadoop does.

Please use https://repo1.maven.org/maven2/org/openlabtesting/protobuf/protoc/2.5.0/ for the aarch64 protoc binary, 
and only override only the groupId used by the protobuf maven plugin from the profile.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   6m 20s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m  0s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  10m 50s |  master passed  |
| +0 |  hbaserecompile  |  22m 11s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 30s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 31s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 19s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   8m 11s |  the patch passed  |
| +0 |  hbaserecompile  |  14m 51s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  4s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   2m 31s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 114m 47s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   2m  9s |  The patch generated 614 ASF License warnings.  |
|  |   | 171m 45s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.PhoenixTTLToolIT |
|   | phoenix.end2end.IndexBuildTimestampIT |
|   | phoenix.end2end.IndexVerificationOldDesignIT |
|   | phoenix.end2end.IndexRepairRegionScannerIT |
|   | phoenix.end2end.IndexExtendedIT |
|   | phoenix.end2end.index.GlobalIndexCheckerIT |
|   | phoenix.end2end.ConcurrentMutationsExtendedIT |
|   | phoenix.end2end.IndexToolForPartialBuildIT |
|   | phoenix.end2end.IndexRebuildTaskIT |
|   | phoenix.end2end.index.IndexMetadataIT |
|   | phoenix.end2end.index.ViewIndexIT |
|   | phoenix.end2end.IndexToolForPartialBuildWithNamespaceEnabledIT |
|   | phoenix.end2end.CsvBulkLoadToolIT |
|   | phoenix.end2end.IndexToolTimeRangeIT |
|   | phoenix.end2end.LogicalTableNameExtendedIT |
|   | phoenix.end2end.LogicalTableNameIT |
|   | phoenix.schema.stats.NamespaceDisabledStatsCollectorIT |
|   | phoenix.end2end.ConcurrentUpsertsWithoutIndexedColsIT |
|   | phoenix.schema.stats.NamespaceEnabledStatsCollectorIT |
|   | phoenix.end2end.BackwardCompatibilityIT |
|   | phoenix.end2end.IndexToolIT |
|   | phoenix.end2end.IndexScrutinyToolIT |
|   | phoenix.end2end.IndexToolForNonTxGlobalIndexIT |
|   | phoenix.end2end.IndexToolForDeleteBeforeRebuildIT |
|   | phoenix.end2end.IndexScrutinyToolForTenantIT |
|   | phoenix.end2end.RegexBulkLoadToolIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1238 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile |
| uname | Linux 23641dc7f3f2 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 3231aad |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/1/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/1/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/1/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 11223 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-pherf . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/1/console |
| versions | git=2.7.4 maven=3.3.9 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by martin-g** [comment]
> Please use https://repo1.maven.org/maven2/org/openlabtesting/protobuf/protoc/2.5.0/ for the aarch64 protoc binary,

I've tried it before choosing os72 but it didn't work because:

1) ${user.home}/.m2/repository/org/openlabtesting/protobuf/protoc/${protobuf-java.version}/protoc-${protobuf-java.version}-${os.detected.classifier}.exe has no executable bit 
```
Caused by: java.io.IOException: Cannot run program "/home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe": error=13, Permission denied
    at java.lang.ProcessBuilder.start (ProcessBuilder.java:1048)
    at java.lang.Runtime.exec (Runtime.java:621)
    at org.codehaus.plexus.util.cli.Commandline.execute (Commandline.java:660)
    at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLineAsCallable (CommandLineUtils.java:136)
    at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLine (CommandLineUtils.java:106)
    at org.codehaus.plexus.util.cli.CommandLineUtils.executeCommandLine (CommandLineUtils.java:89)
    at org.xolstice.maven.plugin.protobuf.Protoc.execute (Protoc.java:242)
    ...
```

2) 2) After "chmod +x ..." it started failing with:
```
[INFO] --- protobuf-maven-plugin:0.6.1:compile (compile-protoc) @ phoenix-core ---
[INFO] Compiling 10 proto file(s) to /home/ubuntu/git/apache/phoenix/phoenix-core/target/generated-sources/protobuf/java
[ERROR] PROTOC FAILED: /home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe: error while loading shared libraries: libprotoc.so.8: cannot open shared object file: No such file or directory

[ERROR] /home/ubuntu/git/apache/phoenix/phoenix-core/src/main/protobuf/PGuidePosts.proto [0:0]: /home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe: error while loading shared libraries: libprotoc.so.8: cannot open shared object file: No such file or directory

[ERROR] /home/ubuntu/git/apache/phoenix/phoenix-core/src/main/protobuf/ServerCacheFactory.proto [0:0]: /home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe: error while loading shared libraries: libprotoc.so.8: cannot open shared object file: No such file or directory
...
```
i.e. the executable is dynamically linked
```
file  /home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe
/home/ubuntu/.m2/repository/org/openlabtesting/protobuf/protoc/2.5.0/protoc-2.5.0-linux-aarch_64.exe: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=ad7c0941f0121611ef1963de1bf6d699e2517c5f, with debug_info, not stripped
```
and nothing provides `libprotoc.so.8` in the LD_LIBRARY_PATH.

I will contact the OpenLab Testing guys and see whether they can release 2.5.0.1 with statically linked executable!

**Comment by stoty** [comment]
Thanks for the explanation and your work on it.

I was suggesting the Openlabtesting version because HBase use(d) that one, and it had 2.5.0 available for arm.

However, if that one has dependency problems, then of course it is not useful for us.

I can see two solutions:

1. You can abandon this change, and specify the compiled ARM protoc binary location in a system property on the maven command line when building.

2. Use the (only ?) known working os72 2.6.1 version of protoc in the aarch64 profile. 
However, in this case please do not change the x86_64 protobuf artifact for 2.5.0, nor the java library versions, only the protoc artifact groupId and version, and only for the arm profile. 
(You'll still need to extract the version and groupId into properties of course)

File: pom.xml (hunk 2 above)
Comment (martin-g) [reply pom.xml]: Reverted to 2.5.0 for both x86_64 and aarch64!
The OpenLabTesting guys released 2.5.0.2 which brings statically linked Protoc.
Here is its build script: https://gist.github.com/liusheng/64aee1b27de037f8b9ccf1873b82c413

**Review by stoty — APPROVED** [review]
+1 I lack the hardware to test this, but LGTM

**Comment by martin-g** [comment]
@stoty about hardware to test it: few other Hadoop related projects use few (3, I think) Linux ARM64 Jenkins agents. See https://ci-hadoop.apache.org/computer/ and look for `arm`. 
Are you interested to test Phoenix on these too ?

**Comment by stoty** [comment]
In theory it'd be nice to test on ARM, but it's not enough of a priority at the moment to put in the work to set up the tests and maintain them.

**Comment by martin-g** [comment]
I could offer my help doing the extra work and maintenance!
Since Phoenix is mainly Java based I don't expect much work related to ARM64. But having CI to notify us when there is a regression would be nice!

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  6s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m  9s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  11m  8s |  master passed  |
| +0 |  hbaserecompile  |  23m 28s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 33s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 12s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   8m  3s |  the patch passed  |
| +0 |  hbaserecompile  |  14m 29s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 29s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 29s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  3s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   2m 13s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 118m 31s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   2m 18s |  The patch generated 614 ASF License warnings.  |
|  |   | 170m 51s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.AuditLoggingIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1238 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile |
| uname | Linux bbf39ec11ce4 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 3231aad |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/2/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/2/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/2/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 11587 (vs. ulimit of 30000) |
| modules | C: phoenix-core . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/2/console |
| versions | git=2.7.4 maven=3.3.9 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   7m  3s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m  1s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  13m 11s |  master passed  |
| +0 |  hbaserecompile  |  25m 20s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 44s |  master passed  |
| +1 :green_heart: |  javadoc  |   2m 31s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 20s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  10m 16s |  the patch passed  |
| +0 |  hbaserecompile  |  17m 45s |  HBase recompiled.  |
| +1 :green_heart: |  compile  |   1m 37s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 37s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  xml  |   0m  2s |  The patch has no ill-formed XML file.  |
| +1 :green_heart: |  javadoc  |   2m 44s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 137m 16s |  root in the patch failed.  |
| -1 :x: |  asflicense  |   1m 40s |  The patch generated 614 ASF License warnings.  |
|  |   | 201m  1s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.BackwardCompatibilityIT |
|   | phoenix.end2end.AuditLoggingIT |
|   | phoenix.end2end.UpsertSelectIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1238 |
| Optional Tests | dupname asflicense javac javadoc unit xml compile |
| uname | Linux 36b2dbc588e9 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | master / 3231aad |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/3/artifact/yetus-general-check/output/patch-unit-root.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/3/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/3/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 7635 (vs. ulimit of 30000) |
| modules | C: phoenix-core . U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1238/3/console |
| versions | git=2.7.4 maven=3.3.9 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.
